### PR TITLE
Sy 1604 clickable manual valves

### PR DESCRIPTION
**Pull request by Lham42**
# Issue Pull Request

## Key Information

- **Linear Issue**: [SY-####](https://linear.app/synnax/issue/)

## Basic Readiness

- [ ] I have performed a self-review of my code.
- [ ] I have added relevant tests to cover the changes to CI.
- [ ] I have added needed QA steps to the [release candidate](/synnaxlabs/synnax/blob/main/.github/PULL_REQUEST_TEMPLATE/rc.md) template that cover these changes.
- [ ] I have updated in-code documentation to reflect the changes.
- [ ] I have updated user-facing documentation to reflect the changes.

### Data Structures

I have ensured that previous versions of stored data structures are properly migrated to new formats in the following projects:

- [ ] Server
- [ ] Console

### API Changes

The following projects have backwards-compatible APIs:

- [ ] Python Client
- [ ] Server
- [ ] TypeScript Client

### CODE REVIEW

File: pluto/src/vis/schematic/Symbols.tsx [hunk 1]
```diff
@@ -264,15 +264,65 @@ export type CompressorProps = ToggleProps<Primitives.CompressorProps>;
 
 // |||||||| STATIC + LABELED ||||||||
 
-export const ReliefValve = createLabeled(Primitives.ReliefValve);
+export const ReliefValve = createLabeled(
+  ({ onChange, ...props }: SymbolProps<ReliefValveProps>): ReactElement => {
+    const [enabled, setEnabled] = useState(props.enabled ?? false);
+
+    const handleClick = () => {
```
Comment (emilbon99) [inline pluto/src/vis/schematic/Symbols.tsx:271]: This code has a lot of duplication. Create a relevant HOC.

**Review by emilbon99 — COMMENTED** [review]
I don't think that all non-interactive symbols should be made interactive by default. The best way to approach this is to add a "dummy" parameter to all toggleable and non-toggleable symbols. If "dummy" is enabled, _then allow toggle_.

**Review by emilbon99 — CHANGES_REQUESTED** [review]
See my comments

**Comment by pjdotson** [comment]
Did you check that this wont break elements on old schematics that don't have the `enabled` prop?

**Comment by Lham42** [comment]
> Did you check that this wont break elements on old schematics that don't have the `enabled` prop?

not yet

File: pluto/src/vis/schematic/registry.ts [hunk 2]
```diff
@@ -444,10 +445,12 @@ const box: Spec<BoxProps> = {
 const reliefValve: Spec<ReliefValveProps> = {
   name: "Relief Valve",
   key: "reliefValve",
-  Form: CommonStyleForm,
+  Form: CommonDummyToggleForm,
   Symbol: ReliefValve,
   defaultProps: (t) => ({
     color: t.colors.gray.l9.rgba255,
+    enabled: false,
```
Comment (pjdotson) [inline pluto/src/vis/schematic/registry.ts:452]: Move these into a `ZERO_DUMMY_TOGGLE_PROPS` constant and use spread notation.

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 3]
```diff
@@ -216,21 +216,30 @@ const Toggle = ({
...
+        enabled && CSS.M("enabled"),
+        triggered && CSS.M("triggered"),
+        className,
+      )}
+      color={Color.cssString(color)}
+      onClick={handleClick}
+      {...props}
+    />
+  );
+};
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:242]: Can revert all of this back

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 4]
```diff
@@ -470,50 +479,69 @@ export const SolenoidValve = ({
   </Toggle>
 );
 
-export interface ReliefValveProps extends DivProps, SVGBasedPrimitiveProps {}
+export interface ReliefValveProps extends ToggleProps, SVGBasedPrimitiveProps {
+  enabled?: boolean;
+  onClick?: MouseEventHandler<HTMLButtonElement>;
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:484]: `enabled` and `onClick` already exist on `ToggleProps`, right?

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 5]
```diff
@@ -470,50 +479,69 @@ export const SolenoidValve = ({
...
+}
 
 export const ReliefValve = ({
   className,
   orientation = "left",
   color,
   scale,
+  enabled = false,
+  onClick,
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:493]: Don't need the `onClick` here, the spread in `...props` covers it

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 6]
```diff
@@ -470,50 +479,69 @@ export const SolenoidValve = ({
...
     >
-      <Path d="M43.5 37L6.35453 18.2035C4.35901 17.1937 2 18.6438 2 20.8803V53.1197C2 55.3562 4.35901 56.8063 6.35453 55.7965L43.5 37ZM43.5 37L80.6455 18.2035C82.641 17.1937 85 18.6438 85 20.8803V53.1197C85 55.3562 82.641 56.8063 80.6455 55.7965L43.5 37Z" />
-      <Path d="M43.5 2 L43.5 37" strokeLinecap="round" />
-      <Path d="M31.8011 14.0802L55.1773 4.29611" strokeLinecap="round" />
-      <Path d="M31.8011 20.0802L55.1773 10.2961" strokeLinecap="round" />
-      <Path d="M31.8011 26.0802L55.1773 16.2961" strokeLinecap="round" />
-    </InternalSVG>
-  </Div>
-);
+      <HandleBoundary orientation={orientation}>
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:504]: Why did you change the width, height, dimensions, etc.?

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 7]
```diff
@@ -819,16 +847,27 @@ export const ISOCap = ({
   </Div>
 );
 
-export interface ManualValveProps extends SVGBasedPrimitiveProps, DivProps {}
+export interface ManualValveProps extends ToggleProps, SVGBasedPrimitiveProps {
+  enabled?: boolean;
+  onClick?: MouseEventHandler<HTMLButtonElement>;
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:852]: Don't need to extend these, see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 8]
```diff
@@ -819,16 +847,27 @@ export const ISOCap = ({
...
+}
 
 export const ManualValve = ({
   className,
   orientation = "left",
   color,
   scale,
+  enabled = false,
+  onClick,
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:861]: don't need the `onClick`, see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 9]
```diff
@@ -819,16 +847,27 @@ export const ISOCap = ({
...
   ...props
 }: ManualValveProps): ReactElement => (
-  <Div className={CSS(CSS.B("manual-valve"), className)} {...props}>
+  <Toggle
+    {...props}
+    orientation={orientation}
+    className={CSS(CSS.B("manual-valve"), className)}
+    enabled={enabled}
+    onClick={onClick}
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:869]: Can remove the `onClick`, covered in `...props`

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 10]
```diff
@@ -1138,16 +1177,27 @@ export const Orifice = ({
   </Div>
 );
 
-export interface NeedleValveProps extends DivProps, SVGBasedPrimitiveProps {}
+export interface NeedleValveProps extends ToggleProps, SVGBasedPrimitiveProps {
+  enabled?: boolean;
+  onClick?: MouseEventHandler<HTMLButtonElement>;
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1182]: see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 11]
```diff
@@ -1138,16 +1177,27 @@ export const Orifice = ({
...
+}
 
 export const NeedleValve = ({
   className,
   orientation = "left",
   color,
   scale,
+  enabled = false,
+  onClick,
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1191]: see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 12]
```diff
@@ -1138,16 +1177,27 @@ export const Orifice = ({
...
   ...props
 }: NeedleValveProps): ReactElement => (
-  <Div className={CSS(CSS.B("needle-valve"), className)} {...props}>
+  <Toggle
+    {...props}
+    orientation={orientation}
+    className={CSS(CSS.B("needle-valve"), className)}
+    enabled={enabled}
+    onClick={onClick}
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1199]: see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 13]
```diff
@@ -1176,53 +1226,56 @@ export const NeedleValve = ({
...
       <Path d="M43.5 21.5L6.35453 2.70349C4.35901 1.69372 2 3.14384 2 5.38029V37.6197C2 39.8562 4.35901 41.3063 6.35453 40.2965L43.5 21.5ZM43.5 21.5L80.6455 2.70349C82.641 1.69372 85 3.14384 85 5.3803V37.6197C85 39.8562 82.641 41.3063 80.6455 40.2965L43.5 21.5Z" />
     </InternalSVG>
-  </Div>
+  </Toggle>
 );
 
-export interface AngledReliefValveProps extends DivProps, SVGBasedPrimitiveProps {}
+export interface AngledReliefValveProps extends ToggleProps, SVGBasedPrimitiveProps {
+  enabled?: boolean;
+  onClick?: MouseEventHandler<HTMLButtonElement>;
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1234]: see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 14]
```diff
@@ -1176,53 +1226,56 @@ export const NeedleValve = ({
...
+}
 
 export const AngledReliefValve = ({
+  color,
   className,
   orientation = "left",
-  color,
   scale,
+  enabled = false,
+  onClick,
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1243]: see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 15]
```diff
@@ -1176,53 +1226,56 @@ export const NeedleValve = ({
...
   ...props
 }: AngledReliefValveProps): ReactElement => (
-  <Div
+  <Toggle
+    {...props}
     orientation={orientation}
     className={CSS(CSS.B("angled-relief-valve"), className)}
-    {...props}
+    enabled={enabled}
+    onClick={onClick}
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1251]: see above

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 16]
```diff
@@ -1176,53 +1226,56 @@ export const NeedleValve = ({
...
   >
     <HandleBoundary orientation={orientation}>
       <Handle
         location="bottom"
         orientation={orientation}
         left={32.8125}
-        top={97.5922}
+        top={97.0278}
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:1258]: Why were the dimensions here changed?

File: pluto/src/vis/schematic/primitives/Primitives.tsx [hunk 17]
```diff
@@ -2383,22 +2436,29 @@ export const Cylinder = ({
 };
 
 export interface SpringLoadedReliefValveProps
-  extends DivProps,
-    SVGBasedPrimitiveProps {}
+  extends ToggleProps,
+    SVGBasedPrimitiveProps {
+  enabled: boolean;
+  onClick?: MouseEventHandler<HTMLButtonElement>;
```
Comment (pjdotson) [inline pluto/src/vis/schematic/primitives/Primitives.tsx:2442]: see above

**Pull request merged**